File: cc/054.cc
// Project Euler Problem 054
//
// How many hands did player one win in the game of poker?

#include <algorithm>
#include <cstdio>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>

#include "myutil.h"

using namespace std;

struct Card {
  int value;
  char suit;
  string name;
};

const string CARD_VALUES = "23456789TJQKA";

// Order stacks of same-valued cards by size, then by value
bool comp(vector<Card> a, vector<Card> b) {
  if (a.size() != b.size()) {
    return a.size() > b.size();
  }
  return a[0].value > b[0].value;
}

class Hand {
public:
  Hand() { };

  void Add(const string& cardStr) {
    Card card = {CARD_VALUES.find(cardStr[0]) + 2, cardStr[1], cardStr};
    // Add to counts
    counts_[card.value]++;
    counts_[1] = counts_[14];  // Aces

    // Add to suits_
    suits_.insert(card.suit);

    // Add to value_stacks_
    vector<vector<Card> >::iterator it2;
    bool inserted = false;
    for (it2=value_stacks_.begin(); it2<value_stacks_.end(); ++it2) {
      if ((*it2)[0].value == card.value) {
	it2->push_back(card);
	inserted = true;
	break;
      }
    }
    if (!inserted) {
      vector<Card> cards;
      cards.push_back(card);
      value_stacks_.push_back(cards);
    }
    sort(value_stacks_.begin(), value_stacks_.end(), comp);
  }

  float Score() {
    // Compute counts summary
    int summary = ValueSummary();
    // Compute score from counts
    float stackScore = 0.0;
    vector<vector<Card> >::const_reverse_iterator rit;
    for (rit=value_stacks_.rbegin(); rit!=value_stacks_.rend(); ++rit) {
      stackScore = (stackScore + (*rit)[0].value) / 15.0;
    }

    float score;
    // Royal Flush
    if ((score = RoyalFlush()) > 0.0) return 9.0 + score;
    // Straight Flush
    if ((score = StraightFlush()) > 0.0) return 8.0 + score;
    // Four of a kind
    if (summary == 41) return 7.0 + stackScore;
    // Full house
    if (summary == 32) return 6.0 + stackScore;
    // Flush
    if (suits_.size() == 1) return 5.0 + stackScore;
    // Straight
    if ((score = Straight()) > 0.0) return 4.0 + score;
    // Three of a kind
    if (summary == 311) return 3.0 + stackScore;
    // Two pairs
    if (summary == 221) return 2.0 + stackScore;
    // One pair
    if (summary == 2111) return 1.0 + stackScore;
    // High card
    return stackScore;
  }

private:
  int ValueSummary() {
    int summary = 0;
    vector<vector<Card> >::const_iterator it;
    for (it=value_stacks_.begin(); it!=value_stacks_.end(); ++it) {
      summary = summary * 10 + it->size();
    }
    return summary;
  }

  float RoyalFlush() {
    if (counts_[14] != 1) return 0.0;  // Ace
    if (counts_[13] != 1) return 0.0;  // King
    return StraightFlush();
  }

  float StraightFlush() {
    if (suits_.size() != 1) return 0.0;
    return Straight();
  }

  float Straight() {
    int inarow = 0;
    for (int i=1; i<=14; ++i) {
      if (counts_[i] == 1) {
	inarow++;
	if (inarow==5) {
	  return i / 15.0;
	}
      } else {
	inarow = 0;
      }
    }
    return 0.0;
  }

private:
  set<char> suits_;
  vector<vector<Card> > value_stacks_;
  map<int, int> counts_;
};

vector<pair<Hand, Hand> > read_hands(const char* filename) {
  vector<pair<Hand, Hand> > retval;
  vector<string> lines = read_lines(filename);
  vector<string>::iterator it;
  for (it=lines.begin(); it != lines.end(); ++it) {
    Hand hand1;
    Hand hand2;
    for (int i=0; i<10; ++i) {
      ((i<5) ? hand1 : hand2).Add((*it).substr(3*i, 2));
    }
    retval.push_back(pair<Hand, Hand>(hand1, hand2));
  }
  return retval;
}

int main(int argc, const char* argv[]) {
  vector<pair<Hand, Hand> > hands = read_hands("../data/p054_poker.txt");
  vector<pair<Hand, Hand> >::iterator it;
  int count = 0;
  for (it=hands.begin(); it != hands.end(); ++it) {
    if ((it->first).Score() > (it->second).Score()) {
      ++count;
    }
  }
  printf("%d\n", count);
}

File: go/problems/054.go
// Project Euler Problem 054
//
// How many hands did player one win in the game of poker?

package problems

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"
)

const VALUES = "23456789TJQKA"
const SUITS = "HDCS"

type Hand struct {
	suits  int   // Base-4 representation of suits, multiples of 11111_4 are flushes
	values int32 // Bitmask of values seen: Ace is (1<<0) and (1<<13)
	cards  []int // Cards seen: units is card value, tens is (# times seen - 1)
}

func (h *Hand) add(cardStr string) {
	value := strings.Index(VALUES, cardStr[:1])
	suit := strings.Index(SUITS, cardStr[1:])

	// Add to values seen
	h.values |= (1 << uint(value+1))
	h.values |= h.values >> 13 // Copy Ace to bit 0

	// Add to suits seen
	h.suits = h.suits*4 + suit

	// Add to cards seen
	for i := range h.cards {
		if h.cards[i]%15 == value {
			h.cards[i] += 15
			return
		}
	}
	h.cards = append(h.cards, value)
}

func (h Hand) valueSummary() (result int) {
	for i := len(h.cards) - 1; i >= 0; i-- {
		result = result*10 + h.cards[i]/15 + 1
	}
	return result
}

func (h Hand) score() (result float32) {
	sort.Ints(h.cards) // Value with most cards last
	summary := h.valueSummary()
	var stackScore float32

	for _, card := range h.cards {
		stackScore = (stackScore + float32(card%15)) / 15
	}

	// Royal flush
	if score := h.royalFlush(); score > 0 {
		return 9 + score
	}
	// Straight flush
	if score := h.straightFlush(); score > 0 {
		return 8 + score
	}
	// Four of a kind
	if summary == 41 {
		return 7 + stackScore
	}
	// Full house
	if summary == 32 {
		return 6 + stackScore
	}
	// Flush
	if h.suits%341 == 0 {
		return 5 + stackScore
	}
	// Straight
	if score := h.straight(); score > 0 {
		return 5 + score
	}
	// Three of a kind
	if summary == 311 {
		return 3 + stackScore
	}
	// Two pairs
	if summary == 221 {
		return 2 + stackScore
	}
	// One pair
	if summary == 2111 {
		return 1 + stackScore
	}
	// High card
	return stackScore
}

func (h Hand) royalFlush() float32 {
	if h.values>>12 != 3 { // Ace, King
		return 0
	}
	return h.straightFlush()
}

func (h Hand) straightFlush() float32 {
	if h.suits%341 > 1 {
		return 0
	}
	return h.straight()
}

func (h Hand) straight() float32 {
	for i := uint(0); i < 10; i++ {
		if (h.values>>i)&0x1f == 0x1f {
			return float32(i) / 15
		}
	}
	return 0
}

func readLines(filename string) (result []string, err error) {
	var file *os.File
	if file, err = os.Open(filename); err != nil {
		return nil, err
	}
	defer file.Close()
	r := bufio.NewReader(file)
	for line, _, err := r.ReadLine(); err != io.EOF; line, _, err = r.ReadLine() {
		if err != nil {
			return nil, err
		}
		result = append(result, string(line))
	}
	return result, err
}

func Problem054() string {
	lines, err := readLines("../data/p054_poker.txt")
	if err != nil {
		return err.Error()
	}
	count := 0
	for _, line := range lines {
		var hand1, hand2 Hand
		cardStrings := strings.Split(line, " ")
		for i := 0; i < 5; i++ {
			hand1.add(cardStrings[i])
			hand2.add(cardStrings[i+5])
		}
		if hand1.score() > hand2.score() {
			count++
		}
	}
	return fmt.Sprint(count)
}

func init() {
	Functions[54] = Problem054
}
